foundations/src/telemetry/log/internal.rs

```diff
@@ -2,93 +2,59 @@ use super::init::LogHarness;
-// Logger and updating the generation number. In general, we try to only read
-// or write the generation number while we hold the appropriate lock on the
-// Logger. It is not mission-critical that we do this. The generation number
-// is just a guard rail to prevent us from going wild. Even if there is some
-// race condition and we're off by one or two, it won't affect its ability to
-// do its job.
-//
-// NOTE: we intentionally use a lock without poisoning here to not
-// panic the threads if they just share telemetry with failed thread.
-pub(crate) type SharedLog = Arc<LogContext>;
-
-#[derive(Debug)]
-pub struct LogContext {
-    // The logger itself. This is the most important part of this struct.
-    pub logger: Arc<parking_lot::RwLock<Logger>>,
-
-    // Generation number, incremented every time the logger is replaced with
-    // a child of itself. Accuracy is not critical as this is only used as a
-    // safety check. Nevertheless, when possible, please be holding the
-    // logger's lock when you read or write the generation number.
-    pub(crate) generation: AtomicU32,
+    // Nesting level, incremented every time the logger is replaced with a child of itself.
+    // Accuracy is not critical as this is only used as a safety check.
+    pub(crate) nesting_level: u32,
 }
 
-impl LogContext {
-    // Create a new LogContext based on a fresh logger. The generation number
-    // is initialized to zero.
+impl LoggerWithNestingTracking {
+    pub const MAX_LOG_NESTING: u32 = 1000;
+    pub const EXCEEDED_MAX_LOG_NESTING_ERROR: &'static str = "foundations: maximum logger nesting exceeded (are add_fields! or set_verbosity being called in a loop?)";
+
+    // Create a new LoggerWithNestingTracking based on a fresh logger. The nesting level is
+    // initialized to zero.
     pub(crate) fn new(logger: Logger) -> Self {
         Self {
-            logger: Arc::new(parking_lot::RwLock::new(logger)),
-            generation: AtomicU32::new(0),
+            inner: logger,
+            nesting_level: 0,
         }
     }
 
-    // Update this LogContext's logger object, and, atomically, increment
-    // the generation number. This mutates 'self' using interior mutability,
-    // even though it only takes a shared reference to self.
-    pub(crate) fn update<F>(&self, f: F)
+    // Update this LoggerWithNestingTracking's logger object, and increment the nesting level.
+    pub(crate) fn update<F>(&mut self, f: F)
     where
         F: FnOnce(&Logger) -> Logger,
     {
-        let mut logger_lock = self.logger.write();
-        let generation = self.generation.fetch_add(1, Ordering::SeqCst) + 1;
-        let logger = f(logger_lock.deref());
-        *logger_lock = logger;
-
-        if generation >= MAX_LOG_GENERATION {
-            panic!("{}", EXCEEDED_MAX_LOG_GENERATION_ERROR);
+        let logger = f(&self.inner);
+        self.inner = logger;
+        self.nesting_level += 1;
+        if self.nesting_level >= Self::MAX_LOG_NESTING {
```

Comment:
Can be `assert!(self.nesting_level < Self::MAX_LOG_NESTING, Self::EXCEEDED_MAX_LOG_NESTING_ERROR)`.